experiment/dash/dash.go

```diff
@@ -3,42 +3,132 @@ package dash
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"time"
 
+	"github.com/montanaflynn/stats"
+	"github.com/neubot/dash/client"
+	neubotModel "github.com/neubot/dash/model"
 	"github.com/ooni/probe-engine/experiment"
 	"github.com/ooni/probe-engine/experiment/handler"
-	"github.com/ooni/probe-engine/experiment/mkevent"
-	"github.com/ooni/probe-engine/measurementkit"
 	"github.com/ooni/probe-engine/model"
 	"github.com/ooni/probe-engine/session"
 )
 
 const (
-	testName    = "dash"
-	testVersion = "0.7.0"
+	testName       = "dash"
+	testVersion    = "0.8.0"
+	defaultTimeout = 55 * time.Second
+	maxRuntime     = 15.0
 )
 
 // Config contains the experiment config.
-type Config struct {
-	// LogLevel is the MK log level. Empty implies "WARNING".
-	LogLevel string
+type Config struct{}
+
+type Simple struct {
+	Connect_latency   float64 `json:"connect_latency"`
+	Median_bitrate    int64   `json:"median_bitrate"`
+	Min_playout_delay float64 `json:"min_playout_delay"`
+}
+
+// TestKeys contains the test keys
+type TestKeys struct {
+	Simple `json:"simple"`
+
+	Client_resolver string `json:"client_resolver"`
+
+	// Failure is the failure string
+	Failure string `json:"failure"`
+
+	Receiver_data []neubotModel.ClientResults `json:"receiver_data"`
 }
 
 func measure(
 	ctx context.Context, sess *session.Session, measurement *model.Measurement,
 	callbacks handler.Callbacks, config Config,
 ) error {
-	settings := measurementkit.NewSettings(
-		"Dash", sess.SoftwareName, sess.SoftwareVersion,
-		sess.CABundlePath(), sess.ProbeASNString(), sess.ProbeCC(),
-		sess.ProbeIP(), sess.ProbeNetworkName(), config.LogLevel,
-	)
-	out, err := measurementkit.StartEx(settings, sess.Logger)
+	testkeys := &TestKeys{
+		Simple: Simple{
+			Connect_latency:   0.0,
+			Min_playout_delay: 0.0,
+		},
+	}
+	measurement.TestKeys = testkeys
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	client := client.New(sess.SoftwareName, sess.SoftwareVersion)
+	channel, err := client.StartDownload(ctx)
 	if err != nil {
+		testkeys.Failure = err.Error()
 		return err
 	}
-	for ev := range out {
-		mkevent.Handle(sess, measurement, ev, callbacks)
+	callbacks.OnProgress(0, fmt.Sprintf("server: %s", client.FQDN))
+	var rates []float64
+	var frame_ready_time float64 = 0.0
+	var play_time float64 = 0.0
+	percentage := 0.00
+	step := 100 / (maxRuntime + 1) / 100
+	for results := range channel {
+		percentage += step
+		message := fmt.Sprintf(
+			"rate: %d kbit/s speed: %.2f kbit/s elapsed: %.2f s",
+			results.Rate,
+			float64(results.Received*8/1024)/results.Elapsed,
+			results.Elapsed,
+		)
+		callbacks.OnProgress(percentage, message)
+		data, err := json.Marshal(results)
+		if err != nil {
+			testkeys.Failure = err.Error()
+			return err
+		}
+		sess.Logger.Debugf("%s", string(data))
+
+		rates = append(rates, float64(results.Rate))
```

Comment:
I was initially skeptical about computing stats inline, but that may actually be better than doing it at the end. Can I suggest to add a comment saying that here we're computing stats inline?

Reply:
To word this differently: I believe the code computing stats is less obvious than the code above. It would therefore be beneficial to introduce a couple of comments says what it is doing and why.